Orange/widgets/unsupervised/owmanifoldlearning.py

```diff
@@ -0,0 +1,284 @@
+        return gui.spin(
+            self.main_area, self, name, minv, maxv, label=label,
+            alignment=Qt.AlignRight,
+            callback=lambda f=self.__spin_parameter_update,
+                            p=name: self.__parameter_changed(f, p))
+
+    def __spin_parameter_update(self, name):
+        self.parameters[name] = getattr(self, name)
+
+    def _create_combo_parameter(self, name, label):
+        self.__combo_parameter_update(name)
+        items = (x[1] for x in getattr(self, name + "_values"))
+        return gui.comboBox(
+            self.main_area, self, name + "_index", label=label, items=items,
+            orientation=Qt.Horizontal,
+            callback=lambda f=self.__combo_parameter_update,
+                            p=name: self.__parameter_changed(f, p))
+
+    def __combo_parameter_update(self, name):
+        index = getattr(self, name + "_index")
+        values = getattr(self, name + "_values")
+        self.parameters[name] = values[index][0]
+
+    def _create_check_parameter(self, name, label):
+        self.__check_parameter_update(name)
+        box = gui.hBox(self.main_area)
+        return gui.checkBox(
+            box, self, name, label,
+            callback=lambda f=self.__check_parameter_update,
+                            p=name: self.__parameter_changed(f, p))
+
+    def __check_parameter_update(self, name):
+        checked = getattr(self, name)
+        values = getattr(self, name + "_values")
+        self.parameters[name] = values[checked]
+
+    def __setattr__(self, name, value):
+        super().__setattr__(name, value)
+        if self.settingsHandler:
+            self.settingsHandler.fast_save(self, name, value)
+        self.settingsHandler.update_defaults(self)
+
+
+class TSNEParametersEditor(ManifoldParametersEditor):
+    _metrics = ("euclidean", "l1", "l2", "manhattan", "cosine",
+                "chebyshev", "jaccard", "mahalanobis", "minkowski",
+                "seuclidean", "sqeuclidean")
+    metric_index = Setting(0)
+    metric_values = [(x, x.capitalize()) for x in _metrics]
+    # init_index = Setting(0)
+    # init_values = (("pca", "PCA"),
+    #                ("random", "Random"))
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.metric_combo = self._create_combo_parameter(
+            "metric", "Metric:")
+        self.parameters["init"] = "pca"
+        # self.init_combo = self._create_combo_parameter(
+        #     "init", "Initialization:")
+
+
+class MDSParametersEditor(ManifoldParametersEditor):
+    max_iter = Setting(300)
+    random_state = Setting(0)
+    random_state_values = (None, 0)
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.max_iter_spin = self._create_spin_parameter(
+            "max_iter", 10, 10 ** 4, "Max iterations:")
+        self.random_state_check = self._create_check_parameter(
+            "random_state", "Replicable (deterministic) learning")
```

Comment:
"Replicable (deterministic) learning" is a bit awkward and too long - does not fit into fixed widget size for me.
Suggestion: invert the option and use "Random initialization" checkbox.
Bonus question: does replicable learning use PCA initialization? because the results look different than in the MDS widget...